cookbooks/tomography_based_plate_motions/plugins/reference_profile.cc

```diff
@@ -0,0 +1,457 @@
+    template <int dim>
+    ReferenceProfile<dim>::ReferenceProfile()
+      :
+      initialized(false),
+      surface_condition_function(2)
+    {}
+
+
+
+    template <int dim>
+    void
+    ReferenceProfile<dim>::update()
+    {
+      if (use_surface_condition_function)
+        {
+          initialized = false;
+          surface_condition_function.set_time(this->get_time());
+          initialize();
+        }
+    }
+
+
+    template <int dim>
+    void
+    ReferenceProfile<dim>::initialize()
+    {
+      if (initialized)
+        return;
+
+      reference_profile.initialize(this->get_mpi_communicator());
+      density_index = reference_profile.get_column_index_from_name("density");
+
+      temperatures.resize(n_points, numbers::signaling_nan<double>());
+      pressures.resize(n_points, numbers::signaling_nan<double>());
+      densities.resize(n_points, numbers::signaling_nan<double>());
+
+      delta_z = this->get_geometry_model().maximal_depth() / (n_points-1);
+
+      MaterialModel::MaterialModelInputs<dim> in(1, this->n_compositional_fields());
+      MaterialModel::MaterialModelOutputs<dim> out(1, this->n_compositional_fields());
+
+      // Constant properties on the reference profile
+      in.strain_rate.resize(0);
+      in.velocity[0] = Tensor <1,dim> ();
+      in.requested_properties = MaterialModel::MaterialProperties::equation_of_state_properties;
+
+      // Check whether gravity is pointing up / out or down / in. In the normal case it should
+      // point down / in and therefore gravity should be positive, leading to increasing
+      // adiabatic pressures and temperatures with depth. In some cases it will point up / out
+      // (e.g. for backward advection), in which case the pressures and temperatures should
+      // decrease with depth and therefore gravity has to be negative in the following equations.
+      const Tensor <1,dim> g = this->get_gravity_model().gravity_vector(this->get_geometry_model().representative_point(0));
+      const Point<dim> point_surf = this->get_geometry_model().representative_point(0);
+      const Point<dim> point_bot = this->get_geometry_model().representative_point(this->get_geometry_model().maximal_depth());
+      const int gravity_direction =  (g * (point_bot - point_surf) >= 0) ?
+                                     1 :
+                                     -1;
+
+      // We want to use different values for the densities in the uppermost vs the lower part of the mantle.
+      double uppermost_mantle_thickness = 0.0;
+      if (Plugins::plugin_type_matches<const MaterialModel::TomographyBasedPlateMotions<dim>>(this->get_material_model()))
+        {
+          const MaterialModel::TomographyBasedPlateMotions<dim> &material_model
+            = Plugins::get_plugin_as_type<const MaterialModel::TomographyBasedPlateMotions<dim>> (this->get_material_model());
+
+          uppermost_mantle_thickness = material_model.get_depth_of_base_of_uppermost_mantle();
+        }
```

Comment:
You initialize `uppermost_mantle_thickness` to zero if that plugin is not found. Is that the correct value? Or are you using the `0.0` value to express that the value is uninitialized and hope that you never use the value below? It seems to me the former, right?

Reply:
Yes, it it the former case. If we use densities from the material model (in this case, `tomography_based_plate_motions`) , we will simply use the densities from the material model until the `uppermost_mantle_thickness` depth. We initialize it to zero so that by default the density profile is from the input reference density profile.